add support to disable auto popup via g:asyncomplete_auto_popup

prabirshrestha opened the pull request:
fixes #5 

@andreypopp let me know if this works for you. You should be able to do something like this.

```vim
let g:asyncomplete_auto_popup = 0

function! s:check_back_space() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~ '\s'
endfunction

inoremap <silent><expr> <TAB>
    \ pumvisible() ? "\<C-n>" :
    \ <SID>check_back_space() ? "\<TAB>" :
    \ asyncomplete#force_refresh()
inoremap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<C-h>"
```

andreypopp commented:
Wow! This is great, thanks!

For some reason it doesn't work for me, I copy pasted relevant pieces from the README:
```
""""" asyncomplete
let g:asyncomplete_auto_popup = 0

function! s:check_back_space() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~ '\s'
endfunction

inoremap <silent><expr> <TAB>
  \ pumvisible() ? "\<C-n>" :
  \ <SID>check_back_space() ? "\<TAB>" :
  \ asyncomplete#force_refresh()
inoremap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<C-h>"

call asyncomplete#register_source(asyncomplete#sources#buffer#get_source_options({
    \ 'name': 'buffer',
    \ 'whitelist': ['*'],
    \ 'completor': function('asyncomplete#sources#buffer#completor'),
    \ }))
```

[My .vimrc entire](https://gist.github.com/ee2dd2851dfceb77ded466182a3ea3fc)

Any ideas how can I debug this?

prabirshrestha commented:
@andreypopp this isn't checked in master yet. So you need to make sure you specifiy the branch name. Let me know if you are still having problems.

```vim
Plug 'prabirshrestha/asyncomplete.vim', { 'branch': 'asyncomplete_auto_popup' }
```

andreypopp commented:
Yes, `Plug '~/Workspace/asyncomplete.vim'` is checked out at this branch. I double checked this and tried 
```
Plug 'prabirshrestha/asyncomplete.vim', { 'branch': 'asyncomplete_auto_popup' }
```
too. No luck.

prabirshrestha commented:
could you try with a minimal vimrc?

andreypopp commented:
Just tried with [this](https://gist.github.com/ed239cac6afbb30b9dc6eff25b791c88).

Note that I'm on nevi's `HEAD`. Going to try with the stable one now.

andreypopp commented:
Hm... neovim 0.1.7 fails with `Unknown function: matchstrpos`

andreypopp commented:
I'm on neovim/neovim#0cd8291 now with this [vimrc](https://gist.github.com/ed239cac6afbb30b9dc6eff25b791c88)... doesn't work.

prabirshrestha commented:
Older version of neovim doesn't work since matchstrpos got checked in around nov 2016. When you try with neovim head make sure to check if the function exists.

prabirshrestha commented:
Another option: Can you modify `autoload/asyncomplete.vim` and add this polyfill and make sure to update `s:python_cm_refresh` function to use `s:matchstrpos` instead of `matchstrpos`. (source: https://github.com/tartansandal/dotfiles/blob/8dbb768658022250118a61370b1d0485a8380ea5/vim/bundle/SimpylFold/autoload/SimpylFold.vim#L85-L95)

```vim
if exists('*matchstrpos')
    function! s:matchstrpos(expr, pattern) abort
        return matchstrpos(a:expr, a:pattern)
    endfunction
else
    function! s:matchstrpos(expr, pattern) abort
        return [matchstr(a:expr, a:pattern), match(a:expr, a:pattern), matchend(a:expr, a:pattern)]
    endfunction
endif

function! s:python_cm_refresh(ctx, force) abort
   ......
   let l:matchpos = s:matchstrpos(l:typed, l:refresh_pattern)
   ....
endfunction
```

andreypopp commented:
Just tried with the 0d44e9b still no luck. Does the minimal vimrc I posted work with your neovim/machine?

andreypopp commented:
Note that no error is occurred — popup just doesn't happen.

prabirshrestha commented:
There is an error but I silently ignore it. You can log it by adding this in vimrc or try..catch

```vim
let g:asyncomplete_log_file = expand('~/asyncomplete.log')
```

or remove this try..catch.

```vim
       try
            call asyncomplete#log('core', 'completor()', l:name, a:ctx)
            call s:sources[l:name].completor(s:sources[l:name], a:ctx)
        catch
            call asyncomplete#log('core', 'notify_sources_to_refresh', 'error', v:exception)
            continue
        endtry
```

I'm seeing the following error. Still looking into why it is throwing error. My vimrc has bunch of other settings enabled. https://gist.github.com/prabirshrestha/279d8b179d9353fe8694#file-vimrc

```
["core","notify_sources_to_refresh","error","Vim(call):E523: Not allowed here"]
````

andreypopp commented:
Strangely I can't see any errors there but popup doesn't appear — [log](https://gist.github.com/9af47d1cf468485c690eb6b2846f9da3)

prabirshrestha commented:
@andreypopp not the best solution but can you try this for now. This seems to work for me and it is what `<Plug>(asyncomplete_force_refresh)` is defined as

```vim
let g:asyncomplete_auto_popup = 0

function! s:check_back_space() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~ '\s'
endfunction

inoremap <silent><expr> <TAB>
  \ pumvisible() ? "\<C-n>" :
  \ <SID>check_back_space() ? "\<TAB>" :
  \ (asyncomplete#menu_selected()?"\<c-y>\<c-r>=asyncomplete#force_refresh()\<CR>":"\<c-r>=asyncomplete#force_refresh()\<CR>")
inoremap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<C-h>"
```

andreypopp commented:
@prabirshrestha that works for me! Awesome! Thank you.

Another thing I noticed is that completion doesn't work before I save the buffer. With `asyncomplete_auto_popup` or without it enabled. After I saved the buffer once — everything works fine. Could be the issues with `asyncomplete-buffer` though.

andreypopp commented:
btw. I really like the feel of asyncomplete after I switched from Python based plugins (deoplete).

prabirshrestha commented:
I'm glad that it is working now. I will leave this branch open for few days and refactor it in couple of days and merge to master when I have more time.

Buffer completion isn't perfect and does have some limitations but it should had worked without saving. Currently known issues are when you open too many files or large file it slows down. My goal is to move to async buffer completion when neovim official supports lua and fallback to vim script if lua doesn't exist. This is why I haven't been trying to spend much effort in buffer completion yet.

I mostly code in typescript these days and have been using it with asyncomplete full time at work for more than 2 months. I have barely needed to use buffer completion. I do play sometimes with go hence added go completion support too. I do have future plans (not any time soon) to add support for rust, flow and ternjs so I get more users to try asyncomplete.

Currently the matcher function isn't good and this is a work I have to do to make it customizable. I tried 
https://github.com/tpope/vim-haystack but was slow so will wait for official lua neovim support. It will default to the current vim script matcher function if lua isn't supported. (Feel free to start one if you want. It should be very easy to write one)

I would recommend you to use `au User asyncomplete_setup` to register your sources so they don't slow down your vim startup. (Will need to update README.md)

```vim
  au User asyncomplete_setup call asyncomplete#register_source(asyncomplete#sources#buffer#get_source_options({
    \ 'name': 'buffer',
    \ 'whitelist': ['*'],
    \ 'completor': function('asyncomplete#sources#buffer#completor'),
    \ }))
```

python setup is a nightmare in vim and I have removed all plugins that use python in my vimrc too and have been using to asyncomplete and fzf currently. Life is good when all you need to care is about your vimrc and vim and not python 😃

prabirshrestha commented:
@andreypopp I have force pushed the branch and refactored it a bit. Could you try it again? I have moved the logic to `asyncomplete#force_refresh` so you should be able to use this now.

```vim
inoremap <silent><expr> <TAB>
  \ pumvisible() ? "\<C-n>" :
  \ <SID>check_back_space() ? "\<TAB>" :
  \ asyncomplete#force_refresh()
```

andreypopp commented:
@prabirshrestha thanks! It works.

prabirshrestha commented:
Thanks for being patience on this issue. I have now merged this in master.

andreypopp commented:
Awesome! Thanks!
